salt/utils/dns.py

```diff
@@ -1,18 +1,563 @@
 # -*- coding: utf-8 -*-
 '''
-Define some generic functions for dns modules
-'''
+Compendium of generic DNS utilities
+
+# Examples:
+dns.lookup(name, rdtype, ...)
+dns.records(name, rdtype, ...)
+
+dns.srv_rec(data)
+dns.srv_data('my1.example.com', 389, prio=10, weight=100)
+dns.srv_name('ldap/tcp', 'example.com')
+
+# can replace
+- my stuff
+    - gai
+    - tmpltools.dns_secure
 
-# Import python libs
-from __future__ import absolute_import
+
+# should replace/augment
+https://blog.widodh.nl/2016/07/calculating-ds-record-from-dnskey-with-python-3/
+https://github.com/saltstack/salt/pull/36180
+- modules:
+    - ddns
+    - dnsutil
+    - dig
+    - libcloud dns
+- states
+    - libcoud dns
+- utils
+    - network
+
+# should fix
+https://github.com/saltstack/salt/issues/25953
+https://github.com/saltstack/salt/issues/20275
+'''
+from __future__ import print_function, absolute_import
+# Python
 import itertools
-import logging
+import socket
+import string
 
-# Import salt libs
-import salt.utils.network
+# Salt
+import salt.modules.cmdmod
+import salt.utils
 from salt._compat import ipaddress
+from salt.utils.odict import OrderedDict #, DefaultOrderedDict
 
+# Debug & Logging
+import logging
+import pprint
+
+# Integrations
+try:
+    import dns.resolver
+    HAS_PYDNS = True
+except ImportError as e:
+    HAS_PYDNS = False
+HAS_DIG = salt.utils.which('dig') is not None
+HAS_HOST = salt.utils.which('host') is not None
+HAS_NSLOOKUP = salt.utils.which('nslookup') is not None
+
```

Comment:
`ISC Bind` has been replaced by `Unbound` in FreeBSD 10, which implies that `nslookup` and `dig` require extra packages to be installed. It would be good if `drill` could be supported too (which is bundled with Unbound and in FreeBSD BASE)